provider/defaultingress/ingress.go

```diff
@@ -0,0 +1,269 @@
+package defaultingress
+
+import (
+	"context"
+	"fmt"
+	"reflect"
+	"strings"
+
+	"github.com/hashicorp/terraform-plugin-framework-validators/listvalidator"
+	"github.com/hashicorp/terraform-plugin-framework/resource/schema"
+	"github.com/hashicorp/terraform-plugin-framework/resource/schema/planmodifier"
+	"github.com/hashicorp/terraform-plugin-framework/resource/schema/stringplanmodifier"
+	"github.com/hashicorp/terraform-plugin-framework/schema/validator"
+	"github.com/hashicorp/terraform-plugin-framework/types"
+	"github.com/hashicorp/terraform-plugin-log/tflog"
+	cmv1 "github.com/openshift-online/ocm-sdk-go/clustersmgmt/v1"
+
+	"github.com/terraform-redhat/terraform-provider-rhcs/provider/common"
+	"github.com/terraform-redhat/terraform-provider-rhcs/provider/common/attrvalidators"
+)
+
+const lowestDefaultIngressVer = "4.14.0-0"
+
+var ValidWildcardPolicies = []string{string(cmv1.WildcardPolicyWildcardsDisallowed),
+	string(cmv1.WildcardPolicyWildcardsAllowed)}
+var DefaultWildcardPolicy = cmv1.WildcardPolicyWildcardsDisallowed
+
+var ValidNamespaceOwnershipPolicies = []string{string(cmv1.NamespaceOwnershipPolicyStrict),
+	string(cmv1.NamespaceOwnershipPolicyInterNamespaceAllowed)}
+var DefaultNamespaceOwnershipPolicy = cmv1.NamespaceOwnershipPolicyStrict
+
+func IngressResource() map[string]schema.Attribute {
+	return map[string]schema.Attribute{
+		"id": schema.StringAttribute{
+			Description: "Unique identifier of the ingress.",
+			Computed:    true,
+			Optional:    true,
+			PlanModifiers: []planmodifier.String{
+				// This passes the state through to the plan, preventing
+				// "known after apply" since we know it won't change.
+				stringplanmodifier.UseStateForUnknown(),
+			},
+		},
+		"route_selectors": schema.MapAttribute{
+			Description: "Route Selectors for ingress. Format should be a comma-separated list of 'key=value'. " +
+				"If no label is specified, all routes will be exposed on both routers." +
+				"For legacy ingress support these are inclusion labels, otherwise they are treated as exclusion label.",
+
+			ElementType: types.StringType,
+			Optional:    true,
+			Validators:  []validator.Map{NotEmptyMapValidator()},
```

Comment:
If this is optional shouldn't it be able to be empty? What does the not empty map validator actually does does?

Reply:
it verifies that if you provide this param it should not be empty map

Reply:
Oh if you provide it, gotcha